### PR TITLE
feat(gatsby-recipes): live stream updates from creating resources

### PR DESCRIPTION
**Pull request by KyleAMathews**
live-updates

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Yeah will update the tests in a bit

File: packages/gatsby-recipes/src/recipe-machine/index.js [hunk 1]
```diff
@@ -205,16 +204,20 @@ const recipeMachine = Machine(
               cb(`TICK`)
             }, 10000)
 
-            applyPlan(context, cb)
-              .then(result => {
+            const emitter = applyPlan(context, cb)
+            emitter.on(`*`, (type, e) => {
+              if (type === `update`) {
+                cb({ type: `onUpdate`, data: e })
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/recipe-machine/index.js:210]: listen for updates from the renderer and send them out

File: packages/gatsby-recipes/src/renderer/render.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import React, { Suspense, useContext, useState } from "react"
-import Queue from "p-queue"
+import Queue from "better-queue"
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:2]: was having odd problems with p-queue not providing events when things finished so moved to the tried & true better-queue

File: packages/gatsby-recipes/src/renderer/render.js [hunk 3]
```diff
@@ -15,11 +16,6 @@ import { useRecipeStep } from "./step-component"
 import { InputProvider } from "./input-provider"
 import { ResourceProvider, useResourceContext } from "./resource-provider"
 
-const queue = new Queue({ concurrency: 5, autoStart: false })
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:18]: with running recipes in parallel — we can't have global vars

File: packages/gatsby-recipes/src/renderer/render.js [hunk 4]
```diff
@@ -202,43 +217,86 @@ const handleResource = (resourceName, context, props) => {
...
-    const result = RecipesReconciler.render(recipeWithWrapper, plan)
-    return transformToPlan(result)
-  } catch (e) {
-    throw e
+  renderResources()
+
+  if (isStream) {
+    return emitter
+  } else {
+    return new Promise((resolve, reject) => {
```
Comment (KyleAMathews) [inline packages/gatsby-recipes/src/renderer/render.js:290]: default to the old promise resolve when things are done & let people opt into the stream interface

File: packages/gatsby-recipes/src/apply-plan.js [hunk 5]
```diff
@@ -1,12 +1,4 @@
...
-  try {
-    // TODO emit updates on every change so user see's progress.
-    const result = await render(context.recipe, cb, context.inputs, true)
-    return result
-  } catch (e) {
-    console.log(e)
-    throw e
-  }
-}
+module.exports = (context, cb) =>
```
Comment (johno) [inline packages/gatsby-recipes/src/apply-plan.js:3]: 😎

File: packages/gatsby-recipes/src/graphql-server/server.js [hunk 6]
```diff
@@ -32,15 +32,19 @@ const pubsub = new PubSub()
 const PORT = process.argv[2] || 50400
 
 let lastState = {}
+let lastDone = 0
 const emitUpdate = state => {
   // eslint-disable-next-line no-unused-vars
   const { lastEvent, ...cleanedState } = state
-  if (!lodash.isEqual(cleanedState, lastState)) {
+  // isEqual doesn't handle values on objects in arrays 🤷‍♀️
+  const newDone = cleanedState.context.plan.filter(r => r.isDone).length
```
Comment (johno) [inline packages/gatsby-recipes/src/graphql-server/server.js:40]: Don't need to do this now, but this block of code might be useful to break out into a compare function next time we're in here so it reads more intuitively when figuring out if state/done has changed.

File: packages/gatsby-recipes/src/renderer/render.js [hunk 7]
```diff
@@ -202,43 +214,86 @@ const handleResource = (resourceName, context, props) => {
...
 
+  const queue = new Queue(
+    async (job, cb) => {
+      const result = await job
+      cb(null, result)
+    },
+    { concurrent: 5 }
+  )
+
+  queue.push(new Promise(resolve => resolve()))
```
Comment (johno) [inline packages/gatsby-recipes/src/renderer/render.js:234]: Not sure I understand pushing this empty promise to the queue. Is this to ensure `isDrained` isn't true until there's an opportunity to begin the render which will queue up resource promises?

Comment (KyleAMathews) [reply packages/gatsby-recipes/src/renderer/render.js:234]: ugh :-( this is testing code left over — I'll remove this + add the compare function

File: packages/gatsby-recipes/src/graphql-server/server.js (hunk 6 above)
Comment (KyleAMathews) [reply packages/gatsby-recipes/src/graphql-server/server.js]: done

**Pull request merged**